content/en/blog/_posts/2023-11-1-Gateway-API-Future.md

```diff
@@ -0,0 +1,90 @@
+---
+layout: blog
+title: "Gateway API Improvements"
+date: 2023-10-19T10:00:00-08:00
+slug: gateway-api-ga
+---
+
+***Authors:*** John Howard (Google), ... TODO
+
+## Backend TLS Policy
```

Comment:
@candita Do you mind adding content here as a comment in the next ~24 hours?

Reply:
`BackendTLSPolicy` is a new Gateway API type used for specifying the TLS configuration of the connection from the Gateway to backend Pods via the Service API object.  It is specified as a Direct PolicyAttachment without defaults or overrides, applied to a Service that accesses a backend, where the BackendTLSPolicy resides in the same namespace as the Service to which it is applied.   One can use the same BackendTLSPolicy for all the different Routes that might point to the referenced Service.

While there were existing ways provided for [TLS to be configured for edge and passthrough termination](https://gateway-api.sigs.k8s.io/guides/tls/?h=tls#tls-configuration), this new API object specifically addresses the configuration of TLS in order to convey HTTPS from the Gateway dataplane to the backend.  This is referred to as "backend TLS termination" and enables the Gateway to know how to connect to a backend Pod that has its own certificate.  

![Termination Types](https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/images/1897-TLStermtypes.png) 

The specification of a BackendTLSPolicy consists of:
- TargetRef - Defines the targeted API object of the policy.  Only Service is allowed.
- TLS - Defines the configuration for TLS, including Hostname, CACertRefs, and WellKnownCACerts.
- Hostname - Defines the Server Name Indication (SNI) that the Gateway uses to connect to the backend.
- CACertRefs - Defines one or more references to objects that contain PEM-encoded TLS certificates,
which are used to establish a TLS handshake between the Gateway and backend.  Either CACertRefs or WellKnownCACerts may be specified, but not both.
- WellKnownCACerts - Specifies whether or not system CA certificates may be used in the TLS
handshake between the Gateway and backend.  Either CACertRefs or WellKnownCACerts may be specified, but not both.

### Examples

#### Using System Certificates

In this example, the `BackendTLSPolicy` is configured to use system certificates to connect with a TLS-encrypted upstream connection where Pods backing the `dev` Service are expected to serve a valid certificate for `dev.example.com`.

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: BackendTLSPolicy
metadata:
  name: tls-upstream-dev
spec:
  targetRef:
    kind: Service
    name: dev-service
    group: ""
  tls:
    wellKnownCACerts: "System"
    hostname: dev.example.com
```

#### Using Explicit CA Certificates

In this example, the `BackendTLSPolicy` is configured to use certificates definedin the configuration map `auth-cert` to connect with a TLS-encrypted upstream connection where Pods backing the `auth` Service are expected to serve a valid certificate for `auth.example.com`.

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: BackendTLSPolicy
metadata:
  name: tls-upstream-auth
spec:
  targetRef:
    kind: Service
    name: auth-service
    group: ""
  tls:
    caCertRefs:
      - kind: ConfigMapReference
        name: auth-cert
        group: ""
    hostname: auth.example.com
```

_(To editor: If there's space, and you want this, use it. It's not really necessary, but who doesn't like pictures)_
The following illustrates a BackendTLSPolicy that configures TLS for a Service serving a backend:
```mermaid
flowchart LR
    client(["client"])
    gateway["Gateway"]
    style gateway fill:#02f,color:#fff
    httproute["HTTP<BR>Route"]
    style httproute fill:#02f,color:#fff
    service["Service"]
    style service fill:#02f,color:#fff
    pod1["Pod"]
    style pod1 fill:#02f,color:#fff
    pod2["Pod"]
    style pod2 fill:#02f,color:#fff
    client -.->|HTTP <br> request| gateway
    gateway --> httproute
    httproute -.->|BackendTLSPolicy|service
    service --> pod1 & pod2
```

Reply:
Need to get this Markdown back? Use GitHub's web UI to “quote reply”, then select all the text and click on the block quote icon. Now you have the Markdown of https://github.com/kubernetes/website/pull/43619#discussion_r1373905020 in your editor. Copy it and cancel the comment.

Reply:
@sftim I'm following @robscott's request to contribute via https://github.com/kubernetes/website/pull/43619/files#r1373535941.  @howardjohn is this what you'd like to work with?

Reply:
This is probably easier:

```suggestion
## Backend TLS Policy

`BackendTLSPolicy` is a new Gateway API type used for specifying the TLS configuration of the connection from the Gateway to backend Pods via the Service API object.  It is specified as a Direct PolicyAttachment without defaults or overrides, applied to a Service that accesses a backend, where the BackendTLSPolicy resides in the same namespace as the Service to which it is applied.   One can use the same BackendTLSPolicy for all the different Routes that might point to the referenced Service.

While there were existing ways provided for [TLS to be configured for edge and passthrough termination](https://gateway-api.sigs.k8s.io/guides/tls/#tls-configuration), this new API object specifically addresses the configuration of TLS in order to convey HTTPS from the Gateway dataplane to the backend.  This is referred to as "backend TLS termination" and enables the Gateway to know how to connect to a backend Pod that has its own certificate.  

![Termination Types](https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/images/1897-TLStermtypes.png) 

The specification of a BackendTLSPolicy consists of:
- TargetRef - Defines the targeted API object of the policy.  Only Service is allowed.
- TLS - Defines the configuration for TLS, including Hostname, CACertRefs, and WellKnownCACerts.
- Hostname - Defines the Server Name Indication (SNI) that the Gateway uses to connect to the backend.
- CACertRefs - Defines one or more references to objects that contain PEM-encoded TLS certificates,
which are used to establish a TLS handshake between the Gateway and backend.  Either CACertRefs or WellKnownCACerts may be specified, but not both.
- WellKnownCACerts - Specifies whether or not system CA certificates may be used in the TLS
handshake between the Gateway and backend.  Either CACertRefs or WellKnownCACerts may be specified, but not both.

### Examples

#### Using System Certificates

In this example, the `BackendTLSPolicy` is configured to use system certificates to connect with a TLS-encrypted upstream connection where Pods backing the `dev` Service are expected to serve a valid certificate for `dev.example.com`.

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: BackendTLSPolicy
metadata:
  name: tls-upstream-dev
spec:
  targetRef:
    kind: Service
    name: dev-service
    group: ""
  tls:
    wellKnownCACerts: "System"
    hostname: dev.example.com
```

#### Using Explicit CA Certificates

In this example, the `BackendTLSPolicy` is configured to use certificates definedin the configuration map `auth-cert` to connect with a TLS-encrypted upstream connection where Pods backing the `auth` Service are expected to serve a valid certificate for `auth.example.com`.

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: BackendTLSPolicy
metadata:
  name: tls-upstream-auth
spec:
  targetRef:
    kind: Service
    name: auth-service
    group: ""
  tls:
    caCertRefs:
      - kind: ConfigMapReference
        name: auth-cert
        group: ""
    hostname: auth.example.com
```

_(To editor: If there's space, and you want this, use it. It's not really necessary, but who doesn't like pictures)_
The following illustrates a BackendTLSPolicy that configures TLS for a Service serving a backend:
```mermaid
flowchart LR
    client(["client"])
    gateway["Gateway"]
    style gateway fill:#02f,color:#fff
    httproute["HTTP<BR>Route"]
    style httproute fill:#02f,color:#fff
    service["Service"]
    style service fill:#02f,color:#fff
    pod1["Pod"]
    style pod1 fill:#02f,color:#fff
    pod2["Pod"]
    style pod2 fill:#02f,color:#fff
    client -.->|HTTP <br> request| gateway
    gateway --> httproute
    httproute -.->|BackendTLSPolicy|service
    service --> pod1 & pod2
```

Reply:
```suggestion
## Backend TLS Policy
## Backend TLS Policy

`BackendTLSPolicy` is a new Gateway API type used for specifying the TLS configuration of the connection from the Gateway to backend Pods via the Service API object.  It is specified as a Direct PolicyAttachment without defaults or overrides, applied to a Service that accesses a backend, where the BackendTLSPolicy resides in the same namespace as the Service to which it is applied.   One can use the same BackendTLSPolicy for all the different Routes that might point to the referenced Service.

While there were existing ways provided for [TLS to be configured for edge and passthrough termination](https://gateway-api.sigs.k8s.io/guides/tls/#tls-configuration), this new API object specifically addresses the configuration of TLS in order to convey HTTPS from the Gateway dataplane to the backend.  This is referred to as "backend TLS termination" and enables the Gateway to know how to connect to a backend Pod that has its own certificate.  

![Termination Types](https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/images/1897-TLStermtypes.png) 

The specification of a BackendTLSPolicy consists of:
- TargetRef - Defines the targeted API object of the policy.  Only Service is allowed.
- TLS - Defines the configuration for TLS, including Hostname, CACertRefs, and WellKnownCACerts.
- Hostname - Defines the Server Name Indication (SNI) that the Gateway uses to connect to the backend.
- CACertRefs - Defines one or more references to objects that contain PEM-encoded TLS certificates,
which are used to establish a TLS handshake between the Gateway and backend.  Either CACertRefs or WellKnownCACerts may be specified, but not both.
- WellKnownCACerts - Specifies whether or not system CA certificates may be used in the TLS
handshake between the Gateway and backend.  Either CACertRefs or WellKnownCACerts may be specified, but not both.

### Examples

#### Using System Certificates

In this example, the `BackendTLSPolicy` is configured to use system certificates to connect with a TLS-encrypted upstream connection where Pods backing the `dev` Service are expected to serve a valid certificate for `dev.example.com`.

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: BackendTLSPolicy
metadata:
  name: tls-upstream-dev
spec:
  targetRef:
    kind: Service
    name: dev-service
    group: ""
  tls:
    wellKnownCACerts: "System"
    hostname: dev.example.com
```

#### Using Explicit CA Certificates

In this example, the `BackendTLSPolicy` is configured to use certificates definedin the configuration map `auth-cert` to connect with a TLS-encrypted upstream connection where Pods backing the `auth` Service are expected to serve a valid certificate for `auth.example.com`.

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: BackendTLSPolicy
metadata:
  name: tls-upstream-auth
spec:
  targetRef:
    kind: Service
    name: auth-service
    group: ""
  tls:
    caCertRefs:
      - kind: ConfigMapReference
        name: auth-cert
        group: ""
    hostname: auth.example.com
```

_(To editor: If there's space, and you want this, use it. It's not really necessary, but who doesn't like pictures)_
The following illustrates a BackendTLSPolicy that configures TLS for a Service serving a backend:
```mermaid
flowchart LR
    client(["client"])
    gateway["Gateway"]
    style gateway fill:#02f,color:#fff
    httproute["HTTP<BR>Route"]
    style httproute fill:#02f,color:#fff
    service["Service"]
    style service fill:#02f,color:#fff
    pod1["Pod"]
    style pod1 fill:#02f,color:#fff
    pod2["Pod"]
    style pod2 fill:#02f,color:#fff
    client -.->|HTTP <br> request| gateway
    gateway --> httproute
    httproute -.->|BackendTLSPolicy|service
    service --> pod1 & pod2
```
```